File: seqspec/seqspec_auth.py
"""Auth module for seqspec CLI."""

import json
from argparse import ArgumentParser, Namespace, RawTextHelpFormatter

from seqspec.auth import AuthProfile, AuthRegistry, init_profile


def setup_auth_args(parser) -> ArgumentParser:
    subparser = parser.add_parser(
        "auth",
        description="""
Manage remote authentication profiles.

Examples:
seqspec auth init --profile igvf --host api.data.igvf.org --host data.igvf.org --username-env IGVF_ACCESS_KEY_ID --password-env IGVF_ACCESS_KEY_SECRET
seqspec auth path
seqspec auth list
seqspec auth resolve https://api.data.igvf.org/reference-files/...
---
""",
        help="Manage remote authentication profiles",
        formatter_class=RawTextHelpFormatter,
    )
    subparsers = subparser.add_subparsers(dest="auth_command", metavar="<AUTH_CMD>")

    init_parser = subparsers.add_parser("init", help="Create or update an auth profile")
    init_parser.add_argument("--profile", required=True)
    init_parser.add_argument("--host", dest="hosts", action="append", required=True)
    init_parser.add_argument("--kind", default="basic", choices=["basic"])
    init_parser.add_argument("--username-env", required=True)
    init_parser.add_argument("--password-env", required=True)

    subparsers.add_parser("path", help="Show auth config path")
    subparsers.add_parser("list", help="List auth profiles")

    resolve_parser = subparsers.add_parser("resolve", help="Resolve profile for a URL")
    resolve_parser.add_argument("url")
    resolve_parser.add_argument("--auth-profile", default=None)

    return subparser


def run_auth(parser: ArgumentParser, args: Namespace) -> None:
    if args.auth_command == "init":
        value = init_profile(
            args.profile,
            AuthProfile(
                hosts=args.hosts,
                kind=args.kind,
                username_env=args.username_env,
                password_env=args.password_env,
            ),
        )
    elif args.auth_command == "path":
        registry = AuthRegistry.load()
        location = registry.location
        value = {
            "kind": "auth_config_path",
            "source": location["source"],
            "path": str(location["path"]) if location["path"] else None,
            "exists": location["exists"],
        }
    elif args.auth_command == "list":
        registry = AuthRegistry.load()
        value = registry.profile_summaries()
    elif args.auth_command == "resolve":
        registry = AuthRegistry.load()
        value = registry.resolve_summary(args.url, args.auth_profile)
    else:
        parser.error("auth requires a subcommand")
        return

    print(json.dumps(value, indent=2))

File: src/seqspec_auth.rs
use crate::auth::{init_profile, AuthKind, AuthProfile, AuthRegistry, RemoteAccess};
use anyhow::Result;
use clap::{Args, Subcommand, ValueEnum};
use serde::Serialize;

#[derive(Debug, Clone, Copy, ValueEnum)]
pub enum OutputFormat {
    Text,
    Json,
}

#[derive(Debug, Args)]
pub struct AuthArgs {
    #[command(subcommand)]
    pub command: AuthCommand,
}

#[derive(Debug, Subcommand)]
pub enum AuthCommand {
    Init(AuthInitArgs),
    Path(AuthPathArgs),
    List(AuthListArgs),
    Resolve(AuthResolveArgs),
}

#[derive(Debug, Args)]
pub struct AuthInitArgs {
    #[arg(long, value_name = "PROFILE", required = true)]
    pub profile: String,

    #[arg(long = "host", value_name = "HOST", required = true)]
    pub hosts: Vec<String>,

    #[arg(long, value_enum, default_value_t = AuthKind::Basic)]
    pub kind: AuthKind,

    #[arg(long, value_name = "ENV", required = true)]
    pub username_env: String,

    #[arg(long, value_name = "ENV", required = true)]
    pub password_env: String,

    #[arg(long, value_enum, default_value_t = OutputFormat::Text)]
    pub format: OutputFormat,
}

#[derive(Debug, Args)]
pub struct AuthPathArgs {
    #[arg(long, value_enum, default_value_t = OutputFormat::Text)]
    pub format: OutputFormat,
}

#[derive(Debug, Args)]
pub struct AuthListArgs {
    #[arg(long, value_enum, default_value_t = OutputFormat::Text)]
    pub format: OutputFormat,
}

#[derive(Debug, Args)]
pub struct AuthResolveArgs {
    #[arg(help = "Remote URL to inspect", value_name = "URL")]
    pub url: String,

    #[arg(long, env = "SEQSPEC_AUTH_PROFILE", value_name = "PROFILE")]
    pub auth_profile: Option<String>,

    #[arg(long, value_enum, default_value_t = OutputFormat::Text)]
    pub format: OutputFormat,
}

#[derive(Debug, Serialize)]
struct PathOutput<'a> {
    kind: &'a str,
    source: &'a str,
    path: Option<String>,
    exists: bool,
}

pub fn run(args: &AuthArgs) -> Result<()> {
    match &args.command {
        AuthCommand::Init(args) => run_init(args),
        AuthCommand::Path(args) => run_path(args),
        AuthCommand::List(args) => run_list(args),
        AuthCommand::Resolve(args) => run_resolve(args),
    }
}

fn run_init(args: &AuthInitArgs) -> Result<()> {
    let output = init_profile(
        &args.profile,
        AuthProfile {
            hosts: args.hosts.clone(),
            kind: args.kind.clone(),
            username_env: args.username_env.clone(),
            password_env: args.password_env.clone(),
        },
    )?;
    print_value(args.format, &output)
}

fn run_path(args: &AuthPathArgs) -> Result<()> {
    let registry = AuthRegistry::load()?;
    let location = registry.location();
    let output = PathOutput {
        kind: "auth_config_path",
        source: &location.source,
        path: location
            .path
            .as_ref()
            .map(|path| path.display().to_string()),
        exists: location.exists,
    };
    print_value(args.format, &output)
}

fn run_list(args: &AuthListArgs) -> Result<()> {
    let registry = AuthRegistry::load()?;
    let profiles = registry.profile_summaries();
    print_value(args.format, &profiles)
}

fn run_resolve(args: &AuthResolveArgs) -> Result<()> {
    let registry = AuthRegistry::load()?;
    let resolved = registry.resolve_summary(&args.url, args.auth_profile.as_deref())?;
    if let Some(profile_name) = args.auth_profile.as_deref() {
        let _ = RemoteAccess::load(Some(profile_name))?;
    }
    print_value(args.format, &resolved)
}

fn print_value<T: Serialize>(format: OutputFormat, value: &T) -> Result<()> {
    match format {
        OutputFormat::Json => {
            println!("{}", serde_json::to_string_pretty(value)?);
        }
        OutputFormat::Text => {
            println!("{}", serde_json::to_string_pretty(value)?);
        }
    }
    Ok(())
}
